File: examples/bun/example04_dx11_clear_window.ts
import {
  DXB_BACKEND_DX11,
  DXB_DEVICE_FLAG_DEBUG,
  DXBRIDGE_NULL_HANDLE,
  DXBridgeLibrary,
  getBooleanArg,
  getStringArg,
  parseArgs,
} from "./dxbridge";
import { Win32Window } from "./win32";

function printUsage(): void {
  console.log(
    "Usage: bun run examples/bun/example04_dx11_clear_window.ts [--dll PATH] [--width 640] [--height 360] [--frames 180] [--color R,G,B,A] [--hidden] [--debug]",
  );
}

function parseRGBA(value: string): [number, number, number, number] {
  const parts = value.split(",").map((part) => Number.parseFloat(part.trim()));
  if (parts.length !== 4 || parts.some((part) => Number.isNaN(part))) {
    throw new Error("--color must be R,G,B,A");
  }
  return [parts[0]!, parts[1]!, parts[2]!, parts[3]!];
}

function main(): void {
  const args = parseArgs(Bun.argv.slice(2));
  if (args.values.help === true) {
    printUsage();
    return;
  }

  const dllPath = getStringArg(args, "dll");
  const width = Number.parseInt(getStringArg(args, "width", "640"), 10);
  const height = Number.parseInt(getStringArg(args, "height", "360"), 10);
  const frames = Number.parseInt(getStringArg(args, "frames", "180"), 10);
  const clearColor = parseRGBA(getStringArg(args, "color", "0.08,0.16,0.32,1.0"));
  const hidden = getBooleanArg(args, "hidden");
  const debug = getBooleanArg(args, "debug");

  const window = new Win32Window("DXBridge Bun Clear Window", width, height, !hidden, "DXBridgeBunClearWindow");
  const dxbridge = new DXBridgeLibrary(dllPath || undefined);

  let device = DXBRIDGE_NULL_HANDLE;
  let swapChain = DXBRIDGE_NULL_HANDLE;
  let rtv = DXBRIDGE_NULL_HANDLE;

  try {
    const hwnd = window.create();
    console.log(`Loaded: ${dxbridge.path}`);
    console.log(`Created HWND: 0x${hwnd.toString(16).toUpperCase().padStart(16, "0")}`);

    dxbridge.init(DXB_BACKEND_DX11);
    device = dxbridge.createDevice({
      backend: DXB_BACKEND_DX11,
      adapterIndex: 0,
      flags: debug ? DXB_DEVICE_FLAG_DEBUG : 0,
    });
    swapChain = dxbridge.createSwapChain(device, {
      hwnd,
      width,
      height,
    });

    const backBuffer = dxbridge.getBackBuffer(swapChain);
    rtv = dxbridge.createRenderTargetView(device, backBuffer);

    console.log("Rendering frames. Close the window manually or wait for auto-close.");
    for (let frame = 0; frame < frames; frame += 1) {
      if (!window.pumpMessages()) {
        break;
      }
      dxbridge.beginFrame(device);
      dxbridge.setRenderTargets(device, rtv);
      dxbridge.clearRenderTarget(rtv, clearColor);
      dxbridge.endFrame(device);
      dxbridge.present(swapChain, 1);
      if (frame + 1 === frames) {
        window.close();
      }
    }

    console.log("Done.");
  } finally {
    if (rtv !== DXBRIDGE_NULL_HANDLE) {
      dxbridge.destroyRenderTargetView(rtv);
    }
    if (swapChain !== DXBRIDGE_NULL_HANDLE) {
      dxbridge.destroySwapChain(swapChain);
    }
    if (device !== DXBRIDGE_NULL_HANDLE) {
      dxbridge.destroyDevice(device);
    }
    dxbridge.shutdown();
    dxbridge.close();
    window.dispose();
  }
}

main();

File: examples/rust/src/bin/example04_dx11_clear_window.rs
#![cfg(windows)]

use dxbridge_rust_examples::cli::{AnyResult, CommandLine};
use dxbridge_rust_examples::dxbridge::{
    DxBridge, DxbDevice, DxbRtv, DxbSwapChain, DXB_BACKEND_DX11, DXB_DEVICE_FLAG_DEBUG,
    DXB_FORMAT_RGBA8_UNORM, DXB_NULL_HANDLE,
};
use dxbridge_rust_examples::win32::Win32Window;

fn main() {
    if let Err(err) = run() {
        eprintln!("{}", err);
        std::process::exit(1);
    }
}

fn run() -> AnyResult<()> {
    let cli = CommandLine::from_env()?;
    if cli.help_requested {
        println!("Usage: example04_dx11_clear_window [--dll PATH] [--width 640] [--height 360] [--frames 180] [--color R,G,B,A] [--hidden] [--debug]");
        return Ok(());
    }

    let width = cli.get_u32("--width", 640)?;
    let height = cli.get_u32("--height", 360)?;
    let frames = cli.get_u32("--frames", 180)?;
    let clear_color = parse_rgba(&cli.get_string("--color", "0.08,0.16,0.32,1.0"))?;

    let mut window = Win32Window::new(
        "DXBridge Rust Clear Window",
        width as i32,
        height as i32,
        !cli.has_flag("--hidden"),
        "DXBridgeRustClearWindow",
    )?;
    let bridge = DxBridge::load(cli.get_optional_string("--dll").as_deref())?;

    let mut initialized = false;
    let mut device: DxbDevice = DXB_NULL_HANDLE;
    let mut swap_chain: DxbSwapChain = DXB_NULL_HANDLE;
    let mut rtv: DxbRtv = DXB_NULL_HANDLE;

    let result = (|| -> AnyResult<()> {
        let hwnd = window.create()?;

        println!("Loaded: {}", bridge.path.display());
        println!("Window: {}", window.title());
        println!("Created HWND: 0x{:016X}", hwnd as usize);

        bridge.init(DXB_BACKEND_DX11)?;
        initialized = true;

        let flags = if cli.has_flag("--debug") {
            DXB_DEVICE_FLAG_DEBUG
        } else {
            0
        };
        device = bridge.create_device(DXB_BACKEND_DX11, 0, flags)?;
        swap_chain = bridge.create_swap_chain(
            device,
            hwnd as *mut _,
            width,
            height,
            DXB_FORMAT_RGBA8_UNORM,
            2,
            0,
        )?;
        let back_buffer = bridge.get_back_buffer(swap_chain)?;
        rtv = bridge.create_render_target_view(device, back_buffer)?;

        println!("Rendering frames. Close the window manually or wait for auto-close.");
        for frame in 0..frames {
            if !window.pump_messages() {
                break;
            }
            bridge.begin_frame(device)?;
            bridge.set_render_targets(device, rtv, DXB_NULL_HANDLE)?;
            bridge.clear_render_target(rtv, clear_color)?;
            bridge.end_frame(device)?;
            bridge.present(swap_chain, 1)?;

            if frame + 1 == frames {
                window.close();
            }
        }

        println!("Done.");
        Ok(())
    })();

    if rtv != DXB_NULL_HANDLE {
        bridge.destroy_render_target_view(rtv);
    }
    if swap_chain != DXB_NULL_HANDLE {
        bridge.destroy_swap_chain(swap_chain);
    }
    if device != DXB_NULL_HANDLE {
        bridge.destroy_device(device);
    }
    if initialized {
        bridge.shutdown();
    }

    result
}

fn parse_rgba(value: &str) -> AnyResult<[f32; 4]> {
    let parts = value.split(',').map(|part| part.trim()).collect::<Vec<_>>();
    if parts.len() != 4 {
        return Err("--color must be R,G,B,A".into());
    }
    let mut rgba = [0.0f32; 4];
    for (index, part) in parts.iter().enumerate() {
        rgba[index] = part.parse::<f32>()?;
    }
    Ok(rgba)
}
